File: 148_sort_list/sortList.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head: return head 
        head = self.mergeSort(head)
        return head

    
    def mergeSort(self, head):
        middle = head
        fast = head

        #find middle node
        while fast.next and fast.next.next:
            fast = fast.next.next
            middle = middle.next
        sec = middle.next #if middle.next else head
        middle.next = None

        #if sec != head.next:
        #print middle.val, sec.val
        if head != middle:
            head = self.mergeSort(head)
            sec = self.mergeSort(sec)
        
        head = self.mergeTwoList(head, sec)
        return head
    
    def mergeTwoList(self, List1, List2):
        #print "a", List1.val, List2.val
        tmp_node = ListNode(-1)
        res = tmp_node
        while List1 and List2:
            if List1.val > List2.val:
                tmp_node.next = List2
                List2 = List2.next
            else:
                tmp_node.next = List1
                List1 = List1.next
            
            tmp_node = tmp_node.next

        if List1:
            while List1:
                tmp_node.next = List1
                List1 = List1.next
                tmp_node = tmp_node.next
        
        if List2:
            while List2:
                tmp_node.next = List2
                List2 = List2.next
                tmp_node = tmp_node.next
        
        #print "b", res.next.val, res.next.next.val
        return res.next

                
    def list2linked(self, List):
        res = ListNode(-1)
        tmp = res 
        for item in List:
            tmp.next = ListNode(item)
            tmp = tmp.next
        
        return res.next


if __name__ == '__main__':
    sol = Solution()
    nums = [] 
    node = sol.list2linked(nums)
     
    node = sol.sortList(node)
    
    while node:
        print node.val
        node = node.next

File: 148_sort_list/sortList.cpp
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x): val(x), next(NULL) {}
};

class Solution {
	public:
    	ListNode* sortList(ListNode* head) {
            /* Divide the list into two part. */
            if(!head || !head->next) return head;
            ListNode* mid = findMiddle(head);
            ListNode* sec = mid -> next;
            mid -> next = NULL;
            /* If middle == head, the list length is 2. */
            if(head != mid) {
                head = sortList(head);
                sec = sortList(sec);
            }
            /* Merge the two lists. */
            return merge(head, sec);
    	}
    private:
        ListNode* findMiddle(ListNode* head) {
            ListNode *slow = head, *fast = head;
            while(fast -> next && fast -> next -> next) {
                slow = slow -> next;
                fast = fast -> next -> next;
            }
            return slow;
        }

        ListNode* merge(ListNode* l1, ListNode* l2) {
            ListNode* res = new ListNode(-1);
            ListNode* temp = res;
            while(l1 && l2) {
                if(l1 -> val > l2 -> val) {
                    temp -> next = l2;
                    l2 = l2 -> next;
                }
                else {
                    temp -> next = l1;
                    l1 = l1 -> next;
                }
                temp = temp -> next;
            }
            
            if(l1) temp -> next = l1;
            if(l2) temp -> next = l2;
            return res -> next;
        }
};

int main() {
    return 0;
}
